plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/WatchManager.java

```diff
@@ -0,0 +1,121 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.openshift.restclient.IOpenShiftWatchListener;
+import com.openshift.restclient.IWatcher;
+import com.openshift.restclient.ResourceKind;
+import com.openshift.restclient.capability.CapabilityVisitor;
+import com.openshift.restclient.capability.resources.IClientCapability;
+import com.openshift.restclient.model.IProject;
+import com.openshift.restclient.model.IResource;
+
+public class WatchManager {
+
+	private static final Collection<String> WATCH_KINDS = Arrays.asList(ResourceKind.BUILD, ResourceKind.POD);
+	private static final String DELIMITER = "|";
+	private static WatchManager instance;
+	private static final Logger LOGGER = LoggerFactory.getLogger(WatchManager.class);
+	
+	private Map<String, IWatcher> watches = Collections.synchronizedMap(new HashMap<>());
+
+	public static WatchManager getInstance() {
+		if(instance == null) {
+			instance = new WatchManager();
+		}
+		return instance;
+	}
+	
+	public void startWatch(final IProject project, final String kind, final List<IResource> resources) {
+		final String projectKind = project.getName() + DELIMITER + kind;
+		if(WATCH_KINDS.contains(kind) && !watches.containsKey(projectKind)){
+			project.accept(new CapabilityVisitor<IClientCapability, IWatcher>() {
+				@Override
+				public IWatcher visit(IClientCapability cap) {
+					final Connection conn = ConnectionsRegistryUtil.getConnectionFor(project);
+					final WatchListener listener = new WatchListener(projectKind, conn, resources);
+					IWatcher watch = cap.getClient().watch(kind, project.getName(), listener);
+					watches.put(projectKind, watch);
+					return watch;
+				}
+			}, null);
+		}
+	}
+	
+	private class WatchListener implements IOpenShiftWatchListener{
+		
+		final private Connection conn;
+		final private List<IResource> resources;
+		final private String key;
+
+		public WatchListener(String key, Connection conn, List<IResource> resources) {
+			this.key = key;
+			this.conn = conn;
+			this.resources = Collections.synchronizedList(resources);
+		}
+		@Override
+		public void connected() {
+			LOGGER.debug("The watch connected to OpenShift connection {}", conn.toString());
+		}
+
+		@Override
+		public void disconnected() {
+			LOGGER.debug("The watch disconnected to OpenShift connection {}", conn.toString());
+			IWatcher watcher = watches.remove(key);
+			watcher.stop();
+		}
+
+		@Override
+		public void error(Throwable err) {
+			LOGGER.error("There was an error watching connection {}",err, conn.toString());
+			IWatcher watcher = watches.remove(key);
+			watcher.stop();
+		}
+
+		@Override
+		public void received(IResource resource, ChangeType change) {
+			LOGGER.debug("Watch received change '{}' for {}", change, resource.toJson());
+			IResource newItem = null;
+			IResource oldItem = null;
+			switch(change) {
+			case ADDED:
+				resources.add(resource);
+				newItem = resource;
+				break;
+			case DELETED:
```

Comment:
better check resources contains the resource in the 1st place. Unlikely to happen but if it does then, bam, indexoutofboundexception in your face